demos/mqtt/subscription-manager/mqtt_subscription_manager.c

```diff
@@ -0,0 +1,441 @@
+
+    while( ( nameIndex < topicNameLength ) && ( filterIndex < topicFilterLength ) )
+    {
+        /* Check if the character in the topic name matches the corresponding
+         * character in the topic filter string. */
+        if( pTopicName[ nameIndex ] == pTopicFilter[ filterIndex ] )
+        {
+            /* Handle special corner cases regarding wildcards at the end of
+             * topic filters, as documented by the MQTT protocol spec. */
+            shouldStopMatching = matchEndWildcards( pTopicFilter,
+                                                    topicFilterLength,
+                                                    topicNameLength,
+                                                    nameIndex,
+                                                    filterIndex );
+            matchFound = shouldStopMatching;
+        }
+        else
+        {
+            /* Check for matching wildcards. */
+            shouldStopMatching = matchWildcards( pTopicFilter,
+                                                 pTopicName,
+                                                 topicNameLength,
+                                                 filterIndex,
+                                                 &nameIndex,
+                                                 &matchFound );
+        }
+
+        if( shouldStopMatching == true )
+        {
+            break;
+        }
+
+        /* Increment indexes. */
+        nameIndex++;
+        filterIndex++;
+    }
+
+    if( matchFound == false )
+    {
+        /* If the end of both strings has been reached, they match. This represents the
+         * case when the topic filter contains a wildcard but neither at the starting nor
+         * the ending. For example, when matching "sport/+/player" topic filter with
+         * "sport/hockey/player" topic name. */
+        matchFound = ( ( nameIndex == topicNameLength ) &&
+                       ( filterIndex == topicFilterLength ) ) ? true : false;
+    }
+
+    return matchFound;
+}
+
+/*-----------------------------------------------------------*/
+
+static bool matchTopic( const char * pTopicName,
+                        const uint16_t topicNameLength,
+                        const char * pTopicFilter,
+                        const uint16_t topicFilterLength )
+{
+    assert( pTopicName != NULL );
+    assert( topicNameLength != 0 );
+    assert( pTopicFilter != NULL );
+    assert( topicFilterLength != 0 );
+
+    bool status = false;
+
+    /* Check for an exact match if the incoming topic name and the registered
+     * topic filter length match. */
+    if( topicNameLength == topicFilterLength )
+    {
+        status = ( strncmp( pTopicName, pTopicFilter, topicNameLength ) == 0 ) ? true : false;
+    }
+
+    /* Match against wildcard characters in topic filter only if the incoming
+     * PUBLISH topic name does not start with a "$" character. */
+    if( ( status == false ) && ( pTopicName[ 0 ] != '$' ) )
```

Comment:
See [Topics beginning with $](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718108). What if the topic name is `$SYS/monitor` and the topic filter is `$SYS/#`. Do we not want to match this as well?

Reply:
Good catch! Have updated the logic to allow wildcard matching for reserved topics when wildcards are at a non-beginning position